Mathlib/Algebra/Module/Injective.lean

```diff
@@ -0,0 +1,467 @@
+#align module.injective_iff_injective_object Module.injective_iff_injective_object
+
+set_option synthInstance.etaExperiment true in
+/-- An `R`-module `Q` satisfies Baer's criterion if any `R`-linear map from an `ideal R` extends to
+an `R`-linear map `R ⟶ Q`-/
+def Module.Baer : Prop :=
+  ∀ (I : Ideal R) (g : I →ₗ[R] Q), ∃ g' : R →ₗ[R] Q, ∀ (x : R) (mem : x ∈ I), g' x = g ⟨x, mem⟩
+set_option linter.uppercaseLean3 false in
+#align module.Baer Module.Baer
+
+namespace Module.Baer
+
+variable {R Q} {M N : Type max u v} [AddCommGroup M] [AddCommGroup N]
+
+variable [Module R M] [Module R N] (i : M →ₗ[R] N) (f : M →ₗ[R] Q)
+
+/-- If we view `M` as a submodule of `N` via the injective linear map `i : M ↪ N`, then a submodule
+between `M` and `N` is a submodule `N'` of `N`. To prove Baer's criterion, we need to consider
+pairs of `(N', f')` such that `M ≤ N' ≤ N` and `f'` extends `f`. -/
+structure ExtensionOf extends LinearPMap R N Q where
+  le : LinearMap.range i ≤ domain
+  is_extension : ∀ m : M, f m = toLinearPMap ⟨i m, le ⟨m, rfl⟩⟩
+set_option linter.uppercaseLean3 false in
+#align module.Baer.extension_of Module.Baer.ExtensionOf
+
+section Ext
+
+variable {i f}
+
+@[ext]
+theorem ExtensionOf.ext {a b : ExtensionOf i f} (domain_eq : a.domain = b.domain)
+    (to_fun_eq :
+      ∀ ⦃x : a.domain⦄ ⦃y : b.domain⦄, (x : N) = y → a.toLinearPMap x = b.toLinearPMap y) :
+    a = b := by
+  rcases a with ⟨a, a_le, e1⟩
+  rcases b with ⟨b, b_le, e2⟩
+  congr
+  exact LinearPMap.ext domain_eq to_fun_eq
+set_option linter.uppercaseLean3 false in
+#align module.Baer.extension_of.ext Module.Baer.ExtensionOf.ext
+
+theorem ExtensionOf.ext_iff {a b : ExtensionOf i f} :
+    a = b ↔ ∃ _ : a.domain = b.domain, ∀ ⦃x : a.domain⦄ ⦃y : b.domain⦄,
+    (x : N) = y → a.toLinearPMap x = b.toLinearPMap y :=
+  ⟨fun r => r ▸ ⟨rfl, fun x y h => congr_arg a.toFun <| by exact_mod_cast h⟩, fun ⟨h1, h2⟩ =>
+    ExtensionOf.ext h1 h2⟩
+set_option linter.uppercaseLean3 false in
+#align module.Baer.extension_of.ext_iff Module.Baer.ExtensionOf.ext_iff
+
+end Ext
+
+instance : Inf (ExtensionOf i f) where
+  inf X1 X2 :=
+    { X1.toLinearPMap ⊓
+        X2.toLinearPMap with
+      le := fun x hx =>
+        (by
+          rcases hx with ⟨x, rfl⟩
+          refine' ⟨X1.le (Set.mem_range_self _), X2.le (Set.mem_range_self _), _⟩
+          rw [← X1.is_extension x, ← X2.is_extension x] :
+          x ∈ X1.toLinearPMap.eqLocus X2.toLinearPMap)
+      is_extension := fun m => X1.is_extension _ }
+
+instance : SemilatticeInf (ExtensionOf i f) :=
+  Function.Injective.semilatticeInf ExtensionOf.toLinearPMap
+    (fun X Y h =>
+      ExtensionOf.ext (by rw [h]) fun x y h' => by
+        have : {x y : N} → (h'' : x = y) → (hx : x ∈ X.toLinearPMap.domain) →
+          (hy : y ∈ Y.toLinearPMap.domain) → X.toLinearPMap ⟨x,hx⟩ = Y.toLinearPMap ⟨y,hy⟩ := by
+            rw [h]
+            intro _ _ h _ _
+            congr
+        apply this h' _ _)
```

Comment:
Maybe add a porting note about the changes you made to the Lean 3 proof? I thought the philosophy was that if you deviate from the script, then leave a porting note.

Reply:
I will once I remember why...

Reply:
Done